### PR TITLE
WIP [bench] CCoinsView(Cache): measure various scenarios

### PR DESCRIPTION
**Pull request by Sjors**
`src/bench/bench_bitcoin -printer=plot -filter=CCoinsView.* > plot.html`:

unknown

Tested scenarios (plus ideas that could be added later):

- [x] access a cached coin
- [x] add 200.000 coins to cache (plot shows average time per coin) 
- [x] flush cache (plot shows average time per coin)
- [ ] load cache from disk
- [ ] access coins that are not cached
- [ ] access coins that are cached but dirty
- [ ] flush dirty coins

The flush test creates a temporary directory on disk.

I normalized the cache access benchmark to take about the same speed as the existing benches do on my machine.

I disabled scaling iterations for the cache addition and flush bench. It doesn't make sense to have more than 1 iteration per eval, because the speed per coin as part of a large flush is a more relevant metric then how often you can flush 1 coin. At the same time I didn't want the benches to use up too much  RAM if someone sets `-scaling` to high number.

To test a bigger cache increase `N_CACHE_SCALE=1`. The cache is about 40 MB by default, try `N_CACHE_SCALE=100` for more realistic scenarios, but note that test needs ~3x RAM.

I'm doing a few things the bench framework doesn't seem designed for. Would like some feedback before I refactor it in the wrong direction:

- [ ] share code with test framework (should support `--disable-test`?)
- [ ] clean up stuff between evals; e.g. I need to reset the coin cache for `CCoinsViewCacheAddCoinFresh` and `CCoinsViewCacheFlush`
- [ ] disable scaling iterations
- [ ] add a memory-scaling argument
- [ ] allow pausing the clock between iterations; e.g. to generate test coins on the fly rather in bulk before the run

### CODE REVIEW

File: src/bench/bench.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ void benchmark::BenchRunner::RunAll(Printer& printer, uint64_t num_evals, double
         if (0 == num_iters) {
             num_iters = 1;
         }
-        State state(p.first, num_evals, num_iters, printer);
+        State state(p.first, num_evals, num_iters, scaling, printer);
```
Comment (MarcoFalke) [inline src/bench/bench.cpp]: Why is this change required? Should be in a separate commit

Comment (Sjors) [reply src/bench/bench.cpp]: I need to disable the effect of the `scaling` setting, see above. The current implementation is a bit of a hack though; once refactored it can go in its own commit.

File: src/bench/ccoins_caching.cpp [hunk 2]
```diff
@@ -6,82 +6,216 @@
...
+SetupDummyCoins(CBasicKeyStore& keystoreRet, CCoinsViewCache& coinsViewCache, int n_coins, int n_keys)
 {
+    std::vector<COutPoint> outpoints;
+    outpoints.resize(n_coins);
+    
+    const std::vector<CKey> keys = SetupDummyKeys(keystoreRet, n_keys);
+    
+    const std::vector<CTransaction> transactions = SetupDummyTransactions(keys, n_coins);
+        
+    for (const CTransaction tx : transactions) {
```
Comment (practicalswift) [inline src/bench/ccoins_caching.cpp]:
```
2018-09-14 12:42:48 pvs-studio-analyzer(pr=13470): bench/ccoins_caching.cpp:80:29: warning: loop variable 'tx' of type 'const CTransaction' creates a copy from type 'const CTransaction' [-Wrange-loop-analysis]
```

Comment (practicalswift) [inline src/bench/ccoins_caching.cpp]: Should be `const CTransaction &`?

File: src/bench/ccoins_caching.cpp [hunk 3]
```diff
@@ -6,82 +6,216 @@
...
+        int64_t nCoinDBCache = 50 * N_CACHE_SCALE;
+        pcoinsdbview.reset(new CCoinsViewDB(nCoinDBCache, false, true));
+        pcoinsTip.reset(new CCoinsViewCache(pcoinsdbview.get()));
+
+        CCoinsViewCache coinsViewCache(pcoinsdbview.get());
+        FastRandomContext rng(true);
+        coinsViewCache.SetBestBlock(rng.rand256());
+                
+        // Add coins to cache:
+        for (const CTransaction tx : transactions) {
```
Comment (practicalswift) [inline src/bench/ccoins_caching.cpp]: Should be `const CTransaction &`?

File: src/bench/bench.h [hunk 4]
```diff
@@ -56,28 +56,59 @@ class State
...
     std::vector<double> m_elapsed_results;
     time_point m_start_time;
 
     bool UpdateTimer(time_point finish_time);
 
-    State(std::string name, uint64_t num_evals, double num_iters, Printer& printer) : m_name(name), m_num_iters_left(0), m_num_iters(num_iters), m_num_evals(num_evals)
+    State(std::string name, uint64_t num_evals, double num_iters, double scaling, Printer& printer) : m_name(name), m_num_iters_left(num_iters), m_num_iters(num_iters), m_scaling(scaling), m_num_evals(num_evals)
     {
     }
+
```
Comment (practicalswift) [inline src/bench/bench.h]:
```
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:70:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:75:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:82:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:88:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:92:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:96:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:99:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
```

File: src/bench/bench.h [hunk 5]
```diff
@@ -56,28 +56,59 @@ class State
...
+    inline bool IsLastIteration()
+    {
+        if (m_num_iters_left == 0) {
+            // Measure time before cleaup
+            bool lastEval = !UpdateTimer(clock::now());
+            if (!lastEval) {
+                m_num_iters_left = m_num_iters;
+            }
+            return true;
+        };
```
Comment (practicalswift) [inline src/bench/bench.h]:
```
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/bench.h:109:  You don't need a ; after a }  [readability/braces] [4]
```

File: src/bench/ccoins_caching.cpp [hunk 6]
```diff
@@ -6,82 +6,216 @@
...
     // Add some keys to the keystore:
-    CKey key[4];
-    for (int i = 0; i < 4; i++) {
-        key[i].MakeNewKey(i % 2);
-        keystoreRet.AddKey(key[i]);
+    for (int i = 0; i < n_keys; i++) {
+        keys[i].MakeNewKey(true);
+        keystoreRet.AddKey(keys[i]);
     }
+
```
Comment (practicalswift) [inline src/bench/ccoins_caching.cpp]:
```
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:38:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:43:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:45:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:47:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:54:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:55:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:58:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:60:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:64:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:73:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
```

Please fix throughout :-)

File: src/bench/ccoins_caching.cpp [hunk 7]
```diff
@@ -6,82 +6,216 @@
...
+    
+    const std::vector<CTransaction> transactions = SetupDummyTransactions(keys, n_txs);
+    
+
+    while (state.IsNewEval()) {
+        // TODO: dedup from test suite
+        SelectParams(CBaseChainParams::REGTEST);
+        // const CChainParams& chainparams = Params();
+        ClearDatadirCache();
+        fs::path pathTemp = fs::temp_directory_path() / "bench_bitcoin" / strprintf("%lu_%i", (unsigned long)GetTime(), (int)(InsecureRandRange(1 << 30)));
```
Comment (practicalswift) [inline src/bench/ccoins_caching.cpp]:
```
2018-09-22 22:13:03 cpplint(pr=13470): src/bench/ccoins_caching.cpp:149:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
```

File: src/bench/bench.h [hunk 8]
```diff
@@ -56,28 +56,59 @@ class State
...
+    uint64_t m_num_iters;
+    double m_scaling;
     const uint64_t m_num_evals;
     std::vector<double> m_elapsed_results;
     time_point m_start_time;
 
     bool UpdateTimer(time_point finish_time);
 
-    State(std::string name, uint64_t num_evals, double num_iters, Printer& printer) : m_name(name), m_num_iters_left(0), m_num_iters(num_iters), m_num_evals(num_evals)
+    State(std::string name, uint64_t num_evals, double num_iters, double scaling, Printer& printer) : m_name(name), m_num_iters_left(num_iters), m_num_iters(num_iters), m_scaling(scaling), m_num_evals(num_evals)
```
Comment (practicalswift) [inline src/bench/bench.h]: `num_iters` should be `uint64_t` and not `double`?

**Comment by Sjors** [comment]
Marking this up for grabs. It should be a matter of rebasing and cleaning up.
